### PR TITLE
Fix #3279 warning closing form

### LINKED ISSUE #3279: warning when closing an encounter form
### Describe the problem
you've entered dictation into the SOAP note in an encounter and mistakenly click the `x` in the tab and close it or mistakenly click `Cancel` at the bottom and close it thus losing all of your work
Screenshot from 2020-03-30 12-56-29

### Describe the solution you'd like

should be warned you are about to lose data entered but not saved

### PR DESCRIPTION
**Pull request by stu01509**
Fixes #3279

#### Short description of what this resolves:
Added warning dialog to show when the user closes the tabs.

image

### CODE REVIEW

**Comment by stu01509** [comment]
Hello @bradymiller @stephenwaite @sjpadgett @tywrenn 

Please review this PR, and give me feedback or suggestion about this dialog style.
Thanks :)

**Comment by sjpadgett** [comment]
Hi Cliff, while the workflow is correct for the alert, we don't want user having to do this for every tab closure. You think providers are stressed now, wait til we give them this:)

These type alerts should happen in forms where if content changed to trigger alert occurs.

**Comment by stu01509** [comment]
Hi @sjpadgett 

Thanks for your reply, If have any request changed Please let me know.

**Comment by sjpadgett** [comment]
Well if I understand @stephenwaite original issue, he was concern that a modified SOAP could be canceled w/o a second chance for user to reconsider. This is common in UX and should somehow extend to all our encounter forms.

I see two places to trigger the alert.
- x-ed out from tab while form is being edited.
- canceled in form while editing.

My first thought is a javascript function, maybe even one placed in our JS utility.js that monitors for content changes and sets a flag that is exposed globally to parent tabs or form(perhaps a session variable) that can be checked by alert before popping up. Set the context for the changes also.

**Comment by stu01509** [comment]
Hi @sjpadgett @stephenwaite and @bradymiller  

I try to in interface/forms/soap/templates/general_new.html write a javascript addEventListener to record text change counter, But it will happen open SOAP page HTTP 500 error.

Also, I've tried in utility.js writer a counter to record, But It will happen redeclaration of let, Any suggestion or advise?

**Comment by stephenwaite** [comment]
hi @stu01509 , not really sure, would you like to push you're best effort so can take a look at the code?

**Comment by stu01509** [comment]
Hi @sjpadgett @stephenwaite and @bradymiller

Finally, I use local storage to record `isSoapEdit`  variables when  SOAP textarea trigger `onkeyup` event. Now the modal only shows when the textarea has modified.

Please review it again, Thanks.

**Comment by stephenwaite** [comment]
nice @stu01509 , that fixes when you click to close the tab but not the cancel button, thank you

**Comment by sjpadgett** [comment]
In form cancel will require you to create a doCancel() function to replace the window.location redirect in the cancel button. The check you onchange flag and act accordingly. Just saying:)

**Comment by stu01509** [comment]
Thanks for @stephenwaite and @sjpadgett review and suggestion,
I will try to improve this solution and push a new commit asap.

File: src/Tabs/TabsWrapper.php [hunk 1]
```diff
@@ -132,9 +131,25 @@ function twSetup(tabsid) {
...
   nav.on("click", "span.icon-close", function() {
-    var panelId = $(this).parent().attr("href").substring(1);
-    top.restoreSession();
-    twCloseTab(tabsid, panelId);
+    const self = $(this);
+    const closeTab = function() {
+        const panelId = self.parent().attr("href").substring(1);
+        top.restoreSession();
+        twCloseTab(tabsid, panelId);
+        localStorage.removeItem('isSoapEdit');
```
Comment (bradymiller) [inline src/Tabs/TabsWrapper.php:139]: shouldn't need to use localStorage. would be more straightforward to just store it in top.* . See here for example of flag that is stored there:
https://github.com/openemr/openemr/blob/master/interface/main/tabs/main.php#L157-L164

Comment (stu01509) [reply src/Tabs/TabsWrapper.php:139]: Wow, Thanks for your instruction, It's very useful to me,
I will try it! 😁

**Comment by stu01509** [comment]
I've tried in `general _new.html` `onkeyup` event to trigger `console.log(jsGlobals) ` variables, But it seems can't access this variable.

image

**Comment by stu01509** [comment]
Now, I still use localstorage to record the `isSoapEdit ` variable, Because of the general _new.html page can't access jsGlobals variable.

Any suggestions or feedback to improve it?

**Comment by bradymiller** [comment]
hi @stu01509 , To store the variable in `top`, you would end up doing something like:
```
top.isSoapEdit = true;
```

**Comment by stu01509** [comment]
Oh my! It's working.
Please review it again 😁

**Comment by stephenwaite** [comment]
hi @stu01509 , the alert works great when you try to close the tab but there's still no alert for clicking the `Cancel` button. Not seeing anything new in the `onclick` attr of that button. Thank you

**Comment by stu01509** [comment]
Hello @stephenwaite 

Sorry, my mistake. Please review it again :)

File: interface/forms/soap/templates/general_new.html [hunk 2]
```diff
@@ -64,5 +64,33 @@ <h2>{xlt t='SOAP'}</h2>
             </div>
         </div>
     </div>
+    {literal}
+    <script>
+        const close = function() {
+            if (top.isSoapEdit === true) {
+                dlgopen('', '', 450, 125, '', '<div class="text-danger">Warning</div>', {
```
Comment (bradymiller) [inline interface/forms/soap/templates/general_new.html:71]: translate:
```
dlgopen('', '', 450, 125, '', '<div class="text-danger">{xla t='Warning'}</div>', {
```

File: interface/forms/soap/templates/general_new.html [hunk 3]
```diff
@@ -64,5 +64,33 @@ <h2>{xlt t='SOAP'}</h2>
...
+                    type: 'Alert',
+                    html: '<p>Do you want to close the tabs?</p>',
```
Comment (bradymiller) [inline interface/forms/soap/templates/general_new.html:73]: translate (using the above xla construct)

File: interface/forms/soap/templates/general_new.html [hunk 4]
```diff
@@ -64,5 +64,33 @@ <h2>{xlt t='SOAP'}</h2>
...
+                    buttons: [
+                        {text: 'Cancel', close: true, style: 'default btn-sm'},
+                        {text: 'Close', close: true, style: 'danger btn-sm', click: closeSoap},
```
Comment (bradymiller) [inline interface/forms/soap/templates/general_new.html:76]: translate above 2 labels (using the above xla construct)

File: src/Tabs/TabsWrapper.php [hunk 5]
```diff
@@ -132,9 +131,32 @@ function twSetup(tabsid) {
...
+        twCloseTab(tabsid, panelId);
+    }
+
+    const closeSoap = function() {
+        top.isSoapEdit = false;
+        closeTab();
+    }
+
+    if (self[0].id === 'SOAP' && top.isSoapEdit === true) {
+        dlgopen('', '', 450, 125, '', '<div class="text-danger">Warning</div>', {
```
Comment (bradymiller) [inline src/Tabs/TabsWrapper.php:147]: translate (note this is different than in above smarty context):
```
dlgopen('', '', 450, 125, '', '<div class="text-danger"><?php echo xla("Warning"); ?></div>', {
```

File: src/Tabs/TabsWrapper.php [hunk 6]
```diff
@@ -132,9 +131,32 @@ function twSetup(tabsid) {
...
+            type: 'Alert',
+            html: '<p>Do you want to close the tabs?</p>',
```
Comment (bradymiller) [inline src/Tabs/TabsWrapper.php:149]: translate (using the above xla function)

File: src/Tabs/TabsWrapper.php [hunk 7]
```diff
@@ -132,9 +131,32 @@ function twSetup(tabsid) {
...
+            buttons: [
+                {text: 'Cancel', close: true, style: 'default btn-sm'},
+                {text: 'Close', close: true, style: 'danger btn-sm', click: closeSoap},
```
Comment (bradymiller) [inline src/Tabs/TabsWrapper.php:152]: translate above 2 labels (using the above xla function)

**Comment by stu01509** [comment]
Hi @bradymiller 

I try to change 'general_new.html'
to `dlgopen('', '', 450, 125, '', '<div class="text-danger">{xla t="Warning"}</div>', {`

But it looks like have some problem, I guessing maybe is {literal} tag effect?

image

File: interface/forms/soap/templates/general_new.html (hunk 2 above)
Comment (bradymiller) [reply interface/forms/soap/templates/general_new.html:71]: yep, smarty makes things interesting; will need to do:
```
dlgopen('', '', 450, 125, '', '<div class="text-danger">{/literal}{xla t='Warning'}{literal}</div>', {
```

File: src/Tabs/TabsWrapper.php (hunk 5 above)
Comment (stu01509) [reply src/Tabs/TabsWrapper.php:147]: Hi @bradymiller 

Here, The title will disappear, I guessing in here it using `<<<EOD` and `EOD`

image

Comment (bradymiller) [reply src/Tabs/TabsWrapper.php:147]: man, EOD...
your gonna have to look up how to call php functions in EOD (there may be examples of this in OpenEMR codebase, but not sure)

Comment (stu01509) [reply src/Tabs/TabsWrapper.php:147]: Sounds like a huge project 🤣, I will try to fix it.
Thank you.

**Comment by stephenwaite** [comment]
think so @stu01509 , try the `{literal}`, https://www.smarty.net/docs/en/language.function.literal.tpl

**Comment by stu01509** [comment]
Hi @stephenwaite 
Thanks for your information, Let me fix translate in EOD problems :)

**Comment by stu01509** [comment]
Hi @bradymiller and @stephenwaite 

Updated this commit, I used variables to save translate content. 
Please review it again, Thanks.

**Comment by stephenwaite** [comment]
hi @stu01509 , nice work! guess this can be extended to other forms which would build on this and be a larger project? would you like to create the parent project and reference this? thank you

File: src/Tabs/TabsWrapper.php [hunk 8]
```diff
@@ -113,9 +113,13 @@ public static function genJavaScript()
         $s = '';
         if (!defined('INCLUDED_TW_ONETIME_JS')) {
             define('INCLUDED_TW_ONETIME_JS', true);
+            $modalTitle = xla("Warning");
+            $modalContent = xla("Do you want to close the tabs");
```
Comment (bradymiller) [inline src/Tabs/TabsWrapper.php:117]: place the punctuation (`?`) within the constant

**Comment by bradymiller** [comment]
hi @stu01509 , What will happen here if a user opens up two SOAP forms at the same time and works on them?

**Comment by stu01509** [comment]
Hello @bradymiller 

if a user opens up two SOAP forms, and just edit one of form when it closes the anyone form, just show modal dialog.

File: src/Tabs/TabsWrapper.php (hunk 8 above)
Comment (stu01509) [reply src/Tabs/TabsWrapper.php]: Done! Please review it again :)

**Comment by stu01509** [comment]
Hello  @stephenwaite 

Sounds good ideas, I would like to build on this project after this PR merged, Can you open a new issue and assign me? Thanks 😁

**Comment by sjpadgett** [comment]
This is why that global js should be array that includes context. This makes life so much easier.
In top:
var FormAction = [{id:'form id'}, {context:'edit'}, {name:'tab name'}, {status:'changed'}];

then universal function to deal.

Just saying one var should handle all edits and just a suggestion from the busy body that is me:)

**Comment by stu01509** [comment]
Hi @sjpadgett 


Thanks for your suggestion 👍, It's a good idea to deal with all the encounter form status, I will plan it in the next feature.

**Comment by sjpadgett** [comment]
> It's a good idea to deal with all the encounter form status

Okay but I collect royalties! $9.99 ought to do it:)

**Comment by stu01509** [comment]
> royalties

Okay, Let me use PR to instead of royalties 😁

**Comment by bradymiller** [comment]
hi @stu01509 , Testing well and bringing into codebase. Great job persevering on this one and agree that a generic mechanism that leverages this (ie. any form can plug into it) would be very nice. Thanks for the improvement! -brady

**Comment by stu01509** [comment]
Hello @bradymiller 

Let me planing a pretty plan for implementation about this feature, and that can easily to use in any form and show the dialog 😃

**Pull request merged**